File: static/progress-tracker.js
/**
 * Progress Tracker Web Component
 * Real-time scenario execution progress with Server-Sent Events
 */

class ProgressTracker extends HTMLElement {
    constructor() {
        super();
        this.sessionId = null;
        this.eventSource = null;
        this.isVisible = false;
        this.progressData = {
            currentStep: null,
            percentage: 0,
            steps: [],
            logs: [],
            isComplete: false,
            hasFailed: false
        };

        // Step definitions with descriptions
        this.stepDefinitions = {
            'initialization': {
                label: 'Initializing',
                description: 'Preparing scenario execution'
            },
            'repository_creation': {
                label: 'Creating Repositories',
                description: 'Setting up GitHub repositories from templates'
            },
            'component_creation': {
                label: 'Creating Components',
                description: 'Mapping repositories to CloudBees components'
            },
            'flag_creation': {
                label: 'Planning Feature Flags',
                description: 'Defining feature flag configurations'
            },
            'environment_creation': {
                label: 'Setting up Environments',
                description: 'Creating CloudBees environments with variables'
            },
            'application_creation': {
                label: 'Creating Applications',
                description: 'Linking components and environments'
            },
            'environment_fm_token_update': {
                label: 'Configuring SDK Keys',
                description: 'Adding feature management tokens'
            },
            'flag_configuration': {
                label: 'Configuring Feature Flags',
                description: 'Setting up flags across environments'
            }
        };
    }

    connectedCallback() {
        this.render();
        this.attachEventListeners();
    }

    disconnectedCallback() {
        this.cleanup();
    }

    render() {
        this.innerHTML = `
            <div class="progress-modal" id="progress-modal" style="display: none;">
                <div class="progress-card">
                    <div class="progress-header">
                        <h3>Executing Scenario</h3>
                        <div class="progress-percentage">
                            <span id="progress-text">0%</span>
                        </div>
                    </div>

                    <div class="progress-bar-container">
                        <div class="progress-bar">
                            <div class="progress-fill" id="progress-fill" style="width: 0%"></div>
                        </div>
                    </div>

                    <div class="progress-steps" id="progress-steps">
                        <!-- Steps will be populated dynamically -->
                    </div>

                    <div class="progress-logs">
                        <div class="logs-header">
                            <h4>Execution Log</h4>
                            <button class="logs-toggle" id="logs-toggle" aria-expanded="false">
                                Show Details
                            </button>
                        </div>
                        <div class="logs-content" id="logs-content" style="display: none;">
                            <div class="logs-list" id="logs-list">
                                <!-- Logs will be populated here -->
                            </div>
                        </div>
                    </div>

                    <div class="progress-actions" id="progress-actions" style="display: none;">
                        <button class="btn primary" onclick="this.closest('progress-tracker').hide()">
                            Close
                        </button>
                    </div>
                </div>
            </div>
        `;
    }

    attachEventListeners() {
        // Toggle logs visibility
        const logsToggle = this.querySelector('#logs-toggle');
        const logsContent = this.querySelector('#logs-content');

        logsToggle.addEventListener('click', () => {
            const isExpanded = logsToggle.getAttribute('aria-expanded') === 'true';
            logsToggle.setAttribute('aria-expanded', !isExpanded);
            logsToggle.textContent = isExpanded ? 'Show Details' : 'Hide Details';
            logsContent.style.display = isExpanded ? 'none' : 'block';
        });

        // Close modal on escape key
        document.addEventListener('keydown', (e) => {
            if (e.key === 'Escape' && this.isVisible) {
                this.hide();
            }
        });

        // Close modal on backdrop click
        const modal = this.querySelector('#progress-modal');
        modal.addEventListener('click', (e) => {
            if (e.target === modal) {
                this.hide();
            }
        });
    }

    show(sessionId) {
        this.sessionId = sessionId;
        this.isVisible = true;
        this.resetProgress();

        const modal = this.querySelector('#progress-modal');
        modal.style.display = 'flex';

        // Start listening for progress events
        this.startProgressStream();
    }

    hide() {
        this.isVisible = false;
        const modal = this.querySelector('#progress-modal');
        modal.style.display = 'none';

        this.cleanup();
    }

    resetProgress() {
        this.progressData = {
            currentStep: null,
            percentage: 0,
            steps: [],
            logs: [],
            isComplete: false,
            hasFailed: false
        };

        // Reset UI
        this.updateProgressBar(0);
        this.querySelector('#progress-steps').innerHTML = '';
        this.querySelector('#logs-list').innerHTML = '';
        this.querySelector('#progress-actions').style.display = 'none';
    }

    startProgressStream() {
        if (!this.sessionId) {
            console.error('No session ID provided for progress stream');
            return;
        }

        // Close existing connection
        this.cleanup();

        console.log(`Starting progress stream for session: ${this.sessionId}`);
        this.addLog(`Connecting to progress stream...`, 'info');

        try {
            this.eventSource = new EventSource(`/api/scenario/${this.sessionId}/progress`);

            this.eventSource.onopen = (event) => {
                console.log('Progress stream connected');
                this.addLog('Connected to progress stream', 'success');
            };

            this.eventSource.onmessage = (event) => {
                try {
                    console.log('Received progress event:', event.data);
                    const progressEvent = JSON.parse(event.data);
                    this.handleProgressEvent(progressEvent);
                } catch (e) {
                    console.error('Failed to parse progress event:', e, event.data);
                    this.addLog(`Parse error: ${e.message}`, 'error');
                }
            };

            this.eventSource.onerror = (error) => {
                console.error('Progress stream error:', error);
                if (this.eventSource.readyState === EventSource.CLOSED) {
                    this.addLog('Connection closed', 'error');
                } else {
                    this.addLog('Connection error - retrying...', 'error');
                }
            };

        } catch (e) {
            console.error('Failed to start progress stream:', e);
            this.addLog(`Failed to connect: ${e.message}`, 'error');
        }
    }

    handleProgressEvent(event) {
        switch (event.event_type) {
            case 'step_started':
                this.startStep(event.step, event.message, event.percentage);
                break;
            case 'step_completed':
                this.completeStep(event.step, event.message, event.percentage);
                break;
            case 'step_failed':
                this.failStep(event.step, event.message);
                break;
            case 'log_message':
                this.addLog(event.message, 'info');
                break;
            case 'scenario_completed':
                this.completeScenario(event.message, event.details);
                break;
            case 'scenario_failed':
                this.failScenario(event.message, event.details);
                break;
        }
    }

    startStep(stepKey, message, percentage) {
        this.progressData.currentStep = stepKey;
        if (percentage !== undefined) {
            this.updateProgressBar(percentage);
        }

        this.updateStepDisplay();
        this.addLog(message, 'info');
    }

    completeStep(stepKey, message, percentage) {
        if (percentage !== undefined) {
            this.updateProgressBar(percentage);
        }

        // Mark step as completed
        if (!this.progressData.steps.includes(stepKey)) {
            this.progressData.steps.push(stepKey);
        }

        this.updateStepDisplay();
        this.addLog(message, 'success');
    }

    failStep(stepKey, message) {
        this.progressData.hasFailed = true;
        this.addLog(message, 'error');
        this.showActions();
    }

    completeScenario(message, details) {
        this.progressData.isComplete = true;
        this.updateProgressBar(100);
        this.addLog(message, 'success');
        this.showActions();
        this.cleanup();
    }

    failScenario(message, details) {
        this.progressData.hasFailed = true;
        this.addLog(message, 'error');
        this.showActions();
        this.cleanup();
    }

    updateProgressBar(percentage) {
        this.progressData.percentage = percentage;

        const progressFill = this.querySelector('#progress-fill');
        const progressText = this.querySelector('#progress-text');

        progressFill.style.width = `${percentage}%`;
        progressText.textContent = `${percentage}%`;
    }

    updateStepDisplay() {
        const stepsContainer = this.querySelector('#progress-steps');
        const steps = Object.keys(this.stepDefinitions);

        stepsContainer.innerHTML = steps.map(stepKey => {
            const step = this.stepDefinitions[stepKey];
            const isCompleted = this.progressData.steps.includes(stepKey);
            const isCurrent = this.progressData.currentStep === stepKey;
            const isFailed = this.progressData.hasFailed && isCurrent;

            let statusClass = '';
            let statusIcon = '';

            if (isFailed) {
                statusClass = 'failed';
                statusIcon = '<svg width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"><line x1="18" y1="6" x2="6" y2="18"/><line x1="6" y1="6" x2="18" y2="18"/></svg>';
            } else if (isCompleted) {
                statusClass = 'completed';
                statusIcon = '<svg width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"><polyline points="20,6 9,17 4,12"/></svg>';
            } else if (isCurrent) {
                statusClass = 'active';
                statusIcon = '<svg class="btn-spinner" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"><path d="M21 12a9 9 0 11-6.219-8.56"/></svg>';
            }

            return `
                <div class="progress-step ${statusClass}">
                    <div class="step-icon">${statusIcon}</div>
                    <div class="step-content">
                        <div class="step-title">${step.label}</div>
                        <div class="step-description">${step.description}</div>
                    </div>
                </div>
            `;
        }).join('');
    }

    addLog(message, type = 'info') {
        this.progressData.logs.push({
            message,
            type,
            timestamp: new Date().toLocaleTimeString()
        });

        const logsList = this.querySelector('#logs-list');
        const logEntry = document.createElement('div');
        logEntry.className = `log-entry ${type}`;
        logEntry.innerHTML = `
            <span class="log-time">${new Date().toLocaleTimeString()}</span>
            <span class="log-message">${message}</span>
        `;

        logsList.appendChild(logEntry);
        logsList.scrollTop = logsList.scrollHeight;
    }

    showActions() {
        this.querySelector('#progress-actions').style.display = 'flex';
    }

    cleanup() {
        if (this.eventSource) {
            this.eventSource.close();
            this.eventSource = null;
        }
    }
}

// Register the custom element
customElements.define('progress-tracker', ProgressTracker);

File: src/progress_tracker.py
"""Real-time progress tracking for scenario execution."""

import asyncio
import json
import logging
from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional

from pydantic import BaseModel

logger = logging.getLogger(__name__)


class ProgressEventType(str, Enum):
    """Types of progress events."""
    STEP_STARTED = "step_started"
    STEP_COMPLETED = "step_completed"
    STEP_FAILED = "step_failed"
    LOG_MESSAGE = "log_message"
    SCENARIO_COMPLETED = "scenario_completed"
    SCENARIO_FAILED = "scenario_failed"


class ProgressStep(str, Enum):
    """Scenario execution steps."""
    INITIALIZATION = "initialization"
    REPOSITORY_CREATION = "repository_creation"
    COMPONENT_CREATION = "component_creation"
    FLAG_CREATION = "flag_creation"
    ENVIRONMENT_CREATION = "environment_creation"
    APPLICATION_CREATION = "application_creation"
    FLAG_CONFIGURATION = "flag_configuration"
    ENVIRONMENT_FM_TOKEN_UPDATE = "environment_fm_token_update"


class ProgressEvent(BaseModel):
    """A progress event in the scenario execution."""
    event_type: ProgressEventType
    step: Optional[ProgressStep] = None
    message: str
    timestamp: str
    percentage: Optional[int] = None
    details: Optional[Dict[str, Any]] = None


class ProgressTracker:
    """Manages real-time progress tracking for scenario execution."""

    def __init__(self, session_id: str):
        self.session_id = session_id
        self.events_queue: asyncio.Queue = asyncio.Queue()
        self.current_step = ProgressStep.INITIALIZATION
        self.step_percentages = {
            ProgressStep.INITIALIZATION: 5,
            ProgressStep.REPOSITORY_CREATION: 20,
            ProgressStep.COMPONENT_CREATION: 35,
            ProgressStep.FLAG_CREATION: 45,
            ProgressStep.ENVIRONMENT_CREATION: 60,
            ProgressStep.APPLICATION_CREATION: 80,
            ProgressStep.ENVIRONMENT_FM_TOKEN_UPDATE: 90,
            ProgressStep.FLAG_CONFIGURATION: 100,
        }
        self.completed_steps = set()

    def _calculate_percentage(self, step: ProgressStep) -> int:
        """Calculate overall completion percentage for a step."""
        return self.step_percentages.get(step, 0)

    def _create_event(
        self,
        event_type: ProgressEventType,
        message: str,
        step: Optional[ProgressStep] = None,
        details: Optional[Dict[str, Any]] = None,
    ) -> ProgressEvent:
        """Create a progress event."""
        percentage = None
        if step and event_type in [ProgressEventType.STEP_STARTED, ProgressEventType.STEP_COMPLETED]:
            percentage = self._calculate_percentage(step)

        return ProgressEvent(
            event_type=event_type,
            step=step,
            message=message,
            timestamp=datetime.utcnow().isoformat(),
            percentage=percentage,
            details=details,
        )

    async def _emit_event(self, event: ProgressEvent) -> None:
        """Emit a progress event to the queue."""
        try:
            await self.events_queue.put(event)
            logger.debug(f"Progress event emitted for {self.session_id}: {event.event_type} - {event.message}")
        except Exception as e:
            logger.error(f"Failed to emit progress event: {e}")

    async def start_step(self, step: ProgressStep, message: str, details: Optional[Dict[str, Any]] = None) -> None:
        """Mark a step as started."""
        self.current_step = step
        event = self._create_event(ProgressEventType.STEP_STARTED, message, step, details)
        await self._emit_event(event)

    async def complete_step(self, step: ProgressStep, message: str, details: Optional[Dict[str, Any]] = None) -> None:
        """Mark a step as completed."""
        self.completed_steps.add(step)
        event = self._create_event(ProgressEventType.STEP_COMPLETED, message, step, details)
        await self._emit_event(event)

    async def fail_step(self, step: ProgressStep, message: str, details: Optional[Dict[str, Any]] = None) -> None:
        """Mark a step as failed."""
        event = self._create_event(ProgressEventType.STEP_FAILED, message, step, details)
        await self._emit_event(event)

    async def log_message(self, message: str, details: Optional[Dict[str, Any]] = None) -> None:
        """Emit a log message."""
        event = self._create_event(ProgressEventType.LOG_MESSAGE, message, self.current_step, details)
        await self._emit_event(event)

    async def complete_scenario(self, message: str, summary: Optional[Dict[str, Any]] = None) -> None:
        """Mark the entire scenario as completed."""
        event = self._create_event(ProgressEventType.SCENARIO_COMPLETED, message, details=summary)
        await self._emit_event(event)

    async def fail_scenario(self, message: str, error_details: Optional[Dict[str, Any]] = None) -> None:
        """Mark the entire scenario as failed."""
        event = self._create_event(ProgressEventType.SCENARIO_FAILED, message, details=error_details)
        await self._emit_event(event)

    async def get_events_stream(self):
        """Generator that yields progress events as they occur."""
        logger.info(f"Starting events stream for session {self.session_id}")

        # Send initial connection event
        initial_event = self._create_event(
            ProgressEventType.LOG_MESSAGE,
            "Progress tracking connected"
        )
        yield f"data: {initial_event.model_dump_json()}\n\n"

        while True:
            try:
                # Wait for next event with timeout
                event = await asyncio.wait_for(self.events_queue.get(), timeout=30.0)

                logger.info(f"Streaming event for {self.session_id}: {event.event_type} - {event.message}")

                # Format as Server-Sent Events
                event_data = event.model_dump_json()
                yield f"data: {event_data}\n\n"

                # Break if scenario is complete or failed
                if event.event_type in [ProgressEventType.SCENARIO_COMPLETED, ProgressEventType.SCENARIO_FAILED]:
                    logger.info(f"Scenario finished for {self.session_id}, ending stream")
                    break

            except asyncio.TimeoutError:
                # Send keep-alive event
                logger.debug(f"Sending keep-alive for {self.session_id}")
                yield "data: {\"event_type\": \"keep_alive\", \"message\": \"Connection active\"}\n\n"
            except Exception as e:
                logger.error(f"Error in events stream for {self.session_id}: {e}")
                error_event = self._create_event(
                    ProgressEventType.SCENARIO_FAILED,
                    f"Progress tracking error: {str(e)}"
                )
                yield f"data: {error_event.model_dump_json()}\n\n"
                break

        logger.info(f"Events stream ended for session {self.session_id}")


# Global registry of progress trackers by session ID
_progress_trackers: Dict[str, ProgressTracker] = {}


def create_progress_tracker(session_id: str) -> ProgressTracker:
    """Create and register a new progress tracker for a session."""
    tracker = ProgressTracker(session_id)
    _progress_trackers[session_id] = tracker
    logger.info(f"Created progress tracker for session {session_id}. Total trackers: {len(_progress_trackers)}")
    return tracker


def get_progress_tracker(session_id: str) -> Optional[ProgressTracker]:
    """Get an existing progress tracker for a session."""
    tracker = _progress_trackers.get(session_id)
    if tracker:
        logger.info(f"Found progress tracker for session {session_id}")
    else:
        logger.warning(f"No progress tracker found for session {session_id}. Available: {list(_progress_trackers.keys())}")
    return tracker


def cleanup_progress_tracker(session_id: str) -> None:
    """Clean up a progress tracker after scenario completion."""
    if session_id in _progress_trackers:
        del _progress_trackers[session_id]
        logger.info(f"Cleaned up progress tracker for session {session_id}")